### PR TITLE
Sub-task: #6429 preload links with actions

### LINKED ISSUE #6429: 3.16 - Preload when enabling and updating WP Rocket, when clearing data, changing themes or permalinks
### Context
As a user, I want WP Rocket to identify internal links on the homepage so that those pages can then be pre-optimized for LCP/ATF.

#### add a Controller class, with a dependency on the `Jobs\Manager` class
- add a method `fetch_links` that will send a request to the homepage, get the HTML, find the first 10 internal links, and send them to the jobs manager
- the number of links is modifiable with a new filter `rocket_atf_warmup_links_number`
- the method will bail-out if the filter to disable the ATF optimization feature is disabled
- add tests

#### add a Subscriber class, with a dependency on the Controller created above
- The subscriber will listen to the following actions: `wp_rocket_upgrade`, `permalink_update`, `switch_theme`
- For each one, a callback method will be used to do the fetch links
- The callback for `wp_rocket_upgrade` should only trigger when upgrading from a version lower than 3.16
- add tests

#### add `WP_Rocket\Engine\Media\AboveTheFold\Activation\Activation` & `WP_Rocket\Engine\Media\AboveTheFold\Activation\ServiceProvider`
- The Activation class will have a dependency on the controller
- add callback on this action: `rocket_after_activation`
- the callback method will call the fetch_links() method from the controller

### Acceptance criteria 
_template to be refined with QA_
- Update the plugin to 3.16 from a previous version. Check the DB against the expected result.
- Perform a fresh install. Check the DB against the expected result.

- Bail out from this process when RUCSS is enabled.

###  Effort estimation:
Effort M

### Additional information

To whoever picks up this issue, it could make sense to divide it in 3 successive PRs:

- This issue: First one to manage the fetch_link method on updatE/activation/install but without sending the results to the SaaS yet -> AC would be to check that on update/activation/install, we find the correct links from the homepage. Instead of sending them to the SaaS yet, they could be outputted as logs for testing purposes.
- https://github.com/wp-media/wp-rocket/issues/6470 : Second one would be to replace the log by actually sending to the SaaS, with the appropriate query strings (see https://github.com/wp-media/wp-rocket/issues/6441)
- https://github.com/wp-media/wp-rocket/issues/6471 : Third one would be to change the request to the SaaS for RUCSS by adding config["optimization_list"]=['rucss'] to the payload.

### PR DESCRIPTION
**Pull request by jeawhanlee**
# Description
Sub task of #6429 that fetches links from homepage.

### User documentation

*Explain how this code impacts users.*

### Technical documentation

*Explain how this code works. Diagram & drawings are welcomed.*

## Type of change
*Delete options that are not relevant.*

- [ ] New feature (non-breaking change which adds functionality).
- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] Enhancement (non-breaking change which improves an existing functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as before).

## New dependencies
*List any new dependencies that are required for this change.*

## Risks
*List possible performance & security issues or risks, and explain how they have been mitigated.*

# Checklists

## Feature validation

- [ ] I validated all the Acceptance Criteria. If possible, provide sreenshots or videos.
- [ ] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [ ] I implemented built-in tests to cover the new/changed code.

## Documentation

- [ ] I prepared the user documentation for the feature/enhancement and shared it in the PR or the GitHub issue.
- [ ] The user documentation covers new/changed entry points (endpoints, WP hooks, configuration files, ...).
- [ ] I prepared the technical documentation if needed, and shared it in the PR or the GitHub issue.

## Code style
- [ ] I wrote self-explanatory code about what it does.
- [ ] I wrote comments to explain why it does it.
- [ ] I named variables and functions explicitely.
- [ ] I protected entry points against unexpected inputs.
- [ ] I did not introduce unecessary complexity.
- [ ] I listed the introduced external dependencies explicitely on the PR.
- [ ] I validated the repo-specific guidelines from CONTRIBUTING.md.

## Observability
- [ ] I handled errors when needed.
- [ ] I wrote user-facing messages that are understandable and provide actionable feedbacks.
- [ ] I prepared ways to observe the implemented system (logs, data, etc.).

## Risks
- [ ] I explicitely mentioned performance risks in the PR.
- [ ] I explicitely mentioned security risks in the PR.

### CODE REVIEW

**Comment by jeawhanlee** [comment]
This will hold the base branch for the next 2 task mentioned here: https://github.com/wp-media/wp-rocket/issues/6429#issuecomment-1969454194

**Comment by jeawhanlee** [comment]
Tests will be added in this last task of this PR.

**Review by Miraeld — APPROVED** [review]

File: inc/Engine/Media/AboveTheFold/WarmUp/Controller.php [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+		$args     = [
+			'user-agent' => 'WP Rocket/Pre-fetch Home Links',
+			'timeout'    => 60,
+		];
+
+		$response = wp_remote_get( $home_url, $args );
+
+		if ( ! is_array( $response ) && is_wp_error( $response ) ) {
+			return [];
+		}
```
Comment (remyperona) [inline inc/Engine/Media/AboveTheFold/WarmUp/Controller.php]: This is already done in `wp_remote_retrieve_body()`.

One thing to check though would be the response code, to make sure we receive a 200 response, with `wp_remote_retrieve_response_code()`.

**Review by remyperona — CHANGES_REQUESTED** [review]

File: inc/Engine/Media/AboveTheFold/WarmUp/Controller.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+				$link_host = wp_parse_url( $link );
+				$site_host = wp_parse_url( $home_url );
+
+				/**
+				 * Check for valid link.
+				 * Check that no WP link.
+				 * Check that no external link.
+				 */
+				return wp_http_validate_url( $link ) && ! strpos( $link, 'wordpress.org' ) && $link_host['host'] === $site_host['host'];
+			}
```
Comment (remyperona) [inline inc/Engine/Media/AboveTheFold/WarmUp/Controller.php]: We should also support relative URLs case

**Review by remyperona — CHANGES_REQUESTED** [review]
I think we should already have tests for what is built in this PR. It will help validate the fetch links part and the possible cases.

**Comment by MathieuLamiot** [comment]
@jeawhanlee I am not sure to see how this PR handles triggering fetch_link when clearing the ATF/LCP data? (see [this discussion](https://github.com/wp-media/wp-rocket/issues/6429#issuecomment-1969936207)).
Is it handled?

Thanks 🙏

**Comment by jeawhanlee** [comment]
> @jeawhanlee I am not sure to see how this PR handles triggering fetch_link when clearing the ATF/LCP data? (see [this discussion](https://github.com/wp-media/wp-rocket/issues/6429#issuecomment-1969936207)). Is it handled?
> 
> Thanks 🙏

@MathieuLamiot Thanks for this, I must have skipped this.

**Review by remyperona — COMMENTED** [review]
Aren't we missing the fetch links on activation part?

File: inc/Engine/Media/AboveTheFold/Activation/Activation.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	 */
+	private $manager;
+
+	/**
+	 * Instantiate class.
+	 *
+	 * @param Controller       $controller Controller instance.
+	 * @param ManagerInterface $manager Above the fold Job Manager.
+	 */
+	public function __construct( Controller $controller, ManagerInterface $manager ) {
```
Comment (Miraeld) [inline inc/Engine/Media/AboveTheFold/Activation/Activation.php]: Above the fold manager doesn't exist anymore as part of https://github.com/wp-media/wp-rocket/issues/6461, does it impact here ?

File: inc/Engine/Media/AboveTheFold/Activation/ServiceProvider.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	 * this service provider must have an alias added
+	 * to this array or it will be ignored.
+	 *
+	 * @var array
+	 */
+	protected $provides = [
+		'atf_query',
+		'atf_context',
+		'warmup_controller',
+		'atf_manager',
```
Comment (Miraeld) [inline inc/Engine/Media/AboveTheFold/Activation/ServiceProvider.php]: `atf_manager` doesn't exist anymore as part of https://github.com/wp-media/wp-rocket/issues/6461

**Review by Miraeld — CHANGES_REQUESTED** [review]
There are some references to the `AboveTheFold Manager` but it's been deleted as part of https://github.com/wp-media/wp-rocket/issues/6461, I was wondering if it would impacts here. 

Also there are some lint errors that could be fixed with `phpcs`.

**Review by remyperona — APPROVED** [review]

**Review by hanna-meda — COMMENTED** [review]
Test Run Results >
[testrail-report-595.pdf](https://github.com/wp-media/wp-rocket/files/15062257/testrail-report-595.pdf)
The failing tests, each have their own GH logged.

**Pull request merged**